guideline: Add flags to enforce some guideline rules

ceolin opened the pull request:
These flags will spot violations to Zephyr's codeguideline. Though as they are currently breaking the build, they are under a `MISRA_SANE` option and are disabled by default. The flags enforce (partially) rules:
9.4, 10.4, 10.5, 10.6, 10.7, 10.8, 16.1 and 16.3

Commit:
coding_guideline: Add optional fallthrough enforcement flag

Add implicit-fallthrough flag when CONFIG_MISRA_SANE is set.
It will require __fallthrough macro or a proper comment tells the
compiler that this happens intentionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
coding_guideline: Add optional flag to check invalid conversions

Generate errors for implicit conversions that may alter a value, that
covers conversions implicit conversions to narrower types, different
signal, different essential types (float -> int), ... From GCC
documentation
(https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Warning-Options.html)

"""
Warn for implicit conversions that may alter a value. This includes
conversions between real and integer, like abs (x) when x is double;
conversions between signed and unsigned, like unsigned ui = -1; and
conversions to smaller types, like sqrtf (M_PI). Do not warn for
explicit casts like abs ((int) x) and ui = (unsigned) -1, or if the
value is not changed by the conversion like in abs (2.0). Warnings
about conversions between signed and unsigned integers can be disabled
by using -Wno-sign-conversion.

For C++, also warn for confusing overload resolution for user-defined
conversions; and conversions that never use a type conversion
operator: conversions to void, the same type, a base class or a
reference to them. Warnings about conversions between signed and
unsigned integers are disabled by default in C++ unless
-Wsign-conversion is explicitly enabled.
"""

This partially cover rules: 10.4, 10.5, 10.6, 10.7 and 10.8

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
coding_guideline: Add optional flag to check override initialization

This enforce Rule 9.4

"""
An element of an object shall not be initialized more than once
"""

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
Humm, will create a new option called code guideline with these options that are breaking the build, since we have a test working with for the current option.

tejlmand commented on cmake/compiler/clang/target_warnings.cmake:
```diff
@@ -109,11 +109,25 @@ endmacro()
 
 macro(toolchain_cc_cpp_warning_error_misra_sane dest_var_name)
-  set_ifndef(${dest_var_name} "-Werror=vla")
+    if (NOT ${dest_var_name})
```
indentation seems wrong.

ceolin replied:
Have changed the way the flags were set an consequently didn't touch it. Would you mind review it again ?

yerabolu reviewed (commented):
is it intended to have these changes in 3 different commits ?

ceolin commented:
> is it intended to have these changes in 3 different commits ?

yes, this could put together, but I think it is better this way. Each flag is addressing different rules.

Commit:
coding_guideline: Add an option to enforce rules

Add a Kconfig option to enable compiler flags that help to enforce
some code guideline rules.

Note: As the current code base is not in compliance with the adopted
code guideline, some rules will generate warnings during the
build. This is intended to help to spot violations.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on Kconfig.zephyr:
```diff
@@ -171,6 +171,12 @@ endmenu
 
 menu "Compiler Options"
 
+config CODEGUIDELINE_CHECK
```
CODING_GUIDELINES at least per

https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html

nashif commented on Kconfig.zephyr:
```diff
@@ -171,6 +171,12 @@ endmenu
 menu "Compiler Options"
 
+config CODEGUIDELINE_CHECK
+	bool "Enforce code guideline rules"
```
ditto

nashif commented on Kconfig.zephyr:
```diff
@@ -171,6 +171,12 @@ endmenu
+config CODEGUIDELINE_CHECK
+	bool "Enforce code guideline rules"
+	help
+	  Use available compiler flags to check code guideline rules during
```
ditto
